File: DataBusAbstraction/go/DataBus.go
/*
Copyright (c) 2018 Intel Corporation.

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
*/

package databus

import (
	"errors"
	"strings"
	"sync"

	"github.com/golang/glog"
)

type dataBusContext interface {
	createContext(map[string]string) error
	startTopic(map[string]string) error
	send(map[string]string, interface{}) error
	receive(map[string]string, string, chan interface{}) error
	stopTopic(string) error
	destroyContext() error
}

type DataBus interface {
	ContextCreate(map[string]string) error
	Publish(map[string]string, interface{}) error
	Subscribe(map[string]string, string, CbType) error
	ContextDestroy() error
}

type topicMeta struct {
	topicType string
	data      chan interface{}
	ev        chan string
}

type dataBus struct {
	busType   string
	direction string
	pubTopics map[string]*topicMeta
	subTopics map[string]*topicMeta
	bus       dataBusContext
	mutex     *sync.Mutex
}

// TODO: A dynamic importing?
var dataBusTypes = map[string]string{"OPCUA": "opcua:", "MQTT": "mqtt:", "NATS": "nats:"}

func errHandler(errMsg string, err *error) {
	if r := recover(); r != nil {
		glog.Errorln(r)
		glog.Errorln(errMsg)
		*err = errors.New(errMsg)
	}
}

func NewDataBus() (db *dataBus, err error) {
	defer errHandler("Couldnt create new databus!!!", &err)
	db = &dataBus{}
	db.pubTopics = map[string]*topicMeta{}
	db.subTopics = map[string]*topicMeta{}
	db.mutex = &sync.Mutex{}
	return
}

func (dbus *dataBus) ContextCreate(contextConfig map[string]string) (err error) {
	defer errHandler("DataBus Context Creation Failed!!!", &err)
	dbus.mutex.Lock()
	defer dbus.mutex.Unlock()
	endPoint := contextConfig["endpoint"]

	switch strings.Split(endPoint, "//")[0] {
	case dataBusTypes["OPCUA"]:
		dbus.busType = dataBusTypes["OPCUA"]
		//TODO: Error check
		dbus.bus, err = newOpcuaInstance()
		if err != nil {
			panic("newOpcuaInstance() Failed!!!")
		}
	case dataBusTypes["MQTT"]:
		dbus.busType = dataBusTypes["MQTT"]
		dbus.bus, err = newMqttInstance()
		if err != nil {
			panic("newMqttInstance() Failed!!!")
		}
	case dataBusTypes["NATS"]:
		dbus.busType = dataBusTypes["NATS"]
		panic("TBD!!!")
	default:
		panic("Unsupported DataBus Type!!!")
	}
	//TODO: Error check
	err = dbus.bus.createContext(contextConfig)
	if err != nil {
		panic("createContext() Failed!!!")
	}
	if contextConfig["direction"] == "PUB" {
		dbus.direction = "PUB"
	}
	if contextConfig["direction"] == "SUB" {
		dbus.direction = "SUB"
	}
	glog.Infoln("DataBus Context Created Successfully")
	return
}

func (dbus *dataBus) Publish(topicConfig map[string]string, msgData interface{}) (err error) {
	defer errHandler("DataBus Publish Failed!!!", &err)
	if strings.Contains(dbus.busType, "opcua") {
		err = dbus.bus.send(topicConfig, msgData)
		if err != nil {
			panic("send() Failed!!!")
		}
		return
	}
	dbus.mutex.Lock()
	defer dbus.mutex.Unlock()

	if dbus.direction == "PUB" {
		if !dbus.checkMsgType(topicConfig["dType"], msgData) {
			panic("Message Type Not supported")
		}
		if _, ok := dbus.pubTopics[topicConfig["name"]]; !ok {
			dbus.bus.startTopic(topicConfig)
			dbus.pubTopics[topicConfig["name"]] = &topicMeta{topicType: topicConfig["dType"], data: nil, ev: nil}
		}
		if dbus.pubTopics[topicConfig["name"]].topicType != topicConfig["dType"] {
			panic("Topic name & type NOT matching")
		}
		err = dbus.bus.send(topicConfig, msgData)
		if err != nil {
			panic("send() Failed!!!")
		}
	}
	return
}

//TODO
type CbType func(topic string, msg interface{})

func worker(topic string, dch chan interface{}, ech chan string, cb CbType) {
	for {
		select {
		case data := <-dch:
			glog.V(1).Infoln("Worker receiving...")
			cb(topic, data)
		case <-ech:
			glog.V(1).Infoln("Worker terminating...")
			return
		}
	}
}

func (dbus *dataBus) Subscribe(topicConfig map[string]string, trig string, cb CbType) (err error) {
	defer errHandler("DataBus Subscription Failed!!!", &err)

	if strings.Contains(dbus.busType, "opcua") {
		if dbus.direction == "SUB" && trig == "START" && cb != nil {
			dch := make(chan interface{})
			ech := make(chan string)
			go worker(topicConfig["name"], dch, ech, cb)
			err = dbus.bus.receive(topicConfig, "START", dch)
			if err != nil {
				panic("receive() Failed!!!")
			}
		}
	} else if dbus.direction == "SUB" && trig == "START" && cb != nil {
		if _, ok := dbus.subTopics[topicConfig["name"]]; !ok {
			dch := make(chan interface{})
			ech := make(chan string)
			go worker(topicConfig["name"], dch, ech, cb)
			dbus.subTopics[topicConfig["name"]] = &topicMeta{topicType: topicConfig["dType"], data: dch, ev: ech}
			err = dbus.bus.receive(topicConfig, "START", dch)
			if err != nil {
				panic("receive() Failed!!!")
			}
		}
	} else if dbus.direction == "SUB" && trig == "STOP" {
		//TODO: opcua integration needs to be done
		err = dbus.bus.receive(topicConfig, "STOP", nil)
		if err != nil {
			panic("receive() Failed!!!")
		}
		//Signal worker to stop
		dbus.subTopics[topicConfig["name"]].ev <- "STOP"
		close(dbus.subTopics[topicConfig["name"]].data)
		close(dbus.subTopics[topicConfig["name"]].ev)
		dbus.bus.stopTopic(topicConfig["name"])
		delete(dbus.subTopics, topicConfig["name"])
	} else {
		panic("Wrong parameters passed for Subscribe()")
	}
	return
}

func (dbus *dataBus) ContextDestroy() (err error) {
	defer errHandler("DataBus Context Termination Failed!!!", &err)
	if !strings.Contains(dbus.busType, "opcua") {
		dbus.mutex.Lock()
		defer dbus.mutex.Unlock()
		//Unsubscribe all existing ones
		for k, v := range dbus.subTopics {
			type tConfig map[string]string
			topicConfig := tConfig{"name": k, "dType": v.topicType}
			dbus.bus.receive(topicConfig, "STOP", nil)
			v.ev <- "STOP"
			close(v.data)
			close(v.ev)
			dbus.bus.stopTopic(k)
			delete(dbus.subTopics, k)
		}
	}
	dbus.bus.destroyContext()
	dbus.direction = ""
	glog.Infoln("DataBus Context Terminated")
	return
}

func (dbus *dataBus) checkMsgType(topicType string, msg interface{}) (ret bool) {
	switch msg.(type) {
	case string:
		if topicType == "string" {
			ret = true
		}
	}
	return
}

File: DataBusAbstraction/py/DataBus.py
"""
Copyright (c) 2018 Intel Corporation.

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
"""

from DataBusOpcua import databOpcua
from queue import Queue
from threading import Thread, Lock
from threading import Event
import logging
from Util.log import configure_logging, LOG_LEVELS

busTypes = {"OPCUA": "opcua:"}


def worker(qu, ev, fn, log):
    '''Worker function to fetch data from queue and trigger cb'''
    logger = log
    while True:
        if ev.is_set():
            logger.info("Worker Done")
            break
        try:
            subDict = qu.get(timeout=2)
            fn(subDict["topic"], subDict["data"])
            qu.task_done()
        except Exception:
            # TODO: pass only for Empty exception
            # logger.error("Exception passed")
            pass
    return

# TODO: This library needs to be architected well to have an
# interface which needs to be implemented by it's sub-classes


class databus:
    '''Creates an instance of databus'''

    def __init__(self, log):
        self.logger = log
        self.busType = None
        self.direction = "NONE"
        self.pubTopics = {}
        self.subTopics = {}
        self.bus = None
        self.mutex = Lock()

    def ContextCreate(self, contextConfig):
        '''Create an underlying messagebus context for the databus
        Arguments:
            contextConfig<dict>: Messagebus params to create the context
                <fields>
                "direction": PUB/SUB/NONE - Mutually exclusive
                "name": context namespace (PUB/SUB context namespaces should
                        match)
                "endpoint": messagebus endpoint address
                    <format> proto://host:port/, proto://host:port/.../
                    <examples>
                    OPCUA -> opcua://0.0.0.0:4840/
                "certFile"   : server/client certificate file
                "privateFile": server/client private key file
                "trustFile"  : ca cert used to sign server/client cert
        Return/Exception: Will raise Exception in case of errors'''

        try:
            self.mutex.acquire()
            endpoint = contextConfig["endpoint"]
            # TODO: Check for unique pub/sub contextName?
            if endpoint.split('//')[0] == busTypes["OPCUA"]:
                self.busType = busTypes["OPCUA"]
                self.bus = databOpcua(self.logger)
                self.logger.info("DataBus type: {}".format(busTypes["OPCUA"]))
            else:
                raise Exception("Not a supported BusType")
            self.logger.info(contextConfig)
            try:
                self.bus.createContext(contextConfig)
            except Exception:
                self.logger.error("{} Failure!!!".format(
                    self.ContextCreate.__name__))
                raise
            if contextConfig["direction"] == "PUB":
                self.direction = "PUB"
            elif contextConfig["direction"] == "SUB":
                self.direction = "SUB"
            else:
                raise Exception("Not a supported BusDirection")
        finally:
            self.mutex.release()

    def Publish(self, topicConfig, data):
        '''Publish data on the databus
        Arguments:
            topicConfig<dict>: Publish topic parameters
                <fields>
                "name": Topic name (in hierarchical form with '/' as delimiter)
                    <example> "root/level1/level2/level3"
                "type": Data type associated with the topic
            data: The message whose type should match the topic data type
        Return/Exception: Will raise Exception in case of errors'''

        if "opcua" in self.busType:
            try:
                self.bus.send(topicConfig, data)
            except Exception:
                self.logger.error("{} Failure!!!".format(
                                 self.Publish.__name__))
                raise

    def Subscribe(self, topicConfigs, topicConfigCount, trig, cb=None):
        '''Subscribe data from the databus
        Arguments:
            topicConfigs<list of dicts>: Subscribe topic parameters
                <dict_fields>
                "name": Topic name
                "dType": Data type associated with the topic
            topicConfigCount: length of topicConfigs list
            trig: START/STOP- Start OR Stop Subscription
            cb: A callback function with data as argument
        Return/Exception: Will raise Exception in case of errors'''
        print("self.busType: ", self.busType)
        if "opcua" in self.busType:
            try:
                if (self.direction == "SUB") and (trig == "START") and \
                   (cb is not None):
                    qu = Queue()
                    ev = Event()
                    ev.clear()
                    th = Thread(target=worker,
                                args=(qu, ev, cb, self.logger))
                    th.deamon = True
                    th.start()
                    self.bus.receive(topicConfigs, topicConfigCount,
                                     "START", qu)
            except Exception:
                self.logger.error("receive {} Failure!!!".format(
                                 self.Subscribe.__name__))
                raise

    def ContextDestroy(self):
        '''Destroys the underlying messagebus context
        It unsubscribe all the existing subscriptions too'''

        try:
            if "opcua" in self.busType:
                self.mutex.acquire()
                try:
                    self.bus.destroyContext()
                    self.bus.direction = " "
                except Exception:
                    self.logger.error("{} Failure!!!".format(
                        self.ContextDestroy.__name__))
                    raise
                finally:
                    self.mutex.release()
        except Exception:
            self.logger.error("{} Failure!!!".format(
                self.ContextDestroy.__name__))
            raise

    def checkMsgType(self, topicType, msgData):
        '''Pvt function'''

        if type(msgData) == str:
            if topicType == "string":
                return True
            else:
                return False
